ttnn/cpp/ttnn/operations/moreh/moreh_linear_backward/moreh_linear_backward.cpp

```diff
@@ -0,0 +1,175 @@
+#include "ttnn/cpp/ttnn/operations/moreh/moreh_matmul/moreh_matmul.hpp"
+#include "ttnn/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.hpp"
+
+namespace ttnn::operations::moreh::moreh_linear_backward {
+
+std::tuple<bool, bool, bool> MorehLinearBackward::get_required_outputs(const std::vector<bool>& are_required_outputs) {
+    if (are_required_outputs.size() != 3) {
+        TT_ASSERT(are_required_outputs.size() == 3, "are_required_outputs size must be 3");
+    }
+
+    return {are_required_outputs[0], are_required_outputs[1], are_required_outputs[2]};
+}
+
+void get_tensor_dim(std::vector<uint32_t>& dim, const tt::tt_metal::LegacyShape& shape) {
+    const auto rank = shape.rank();
+    for (auto i = 0; i < rank; ++i) {
+        auto idx = rank - 1 - i;
+
+        // last 2-dim
+        if (idx == rank - 1 || idx == rank - 2) {
+            dim[i] = shape[idx] / tt::constants::TILE_HEIGHT;
+        } else {
+            dim[i] = shape[idx];
+        }
+    }
+
+    log_debug(tt::LogOp, "rank {}", rank);
+    for (auto i = 0; i < tt::tt_metal::MAX_NUM_DIMENSIONS; ++i) {
+        log_debug(tt::LogOp, "dim[{}] = {}", i, dim[i]);
+    }
+}
+
+inline void moreh_linear_backward_validate(
+    const Tensor& output_grad,
+    const Tensor& input,
+    const Tensor& weight,
+    const std::optional<const Tensor>& input_grad,
+    const std::optional<const Tensor>& weight_grad,
+    const std::optional<const Tensor>& bias_grad) {
+    if (input_grad.has_value()) {
+        const auto& input_grad_tensor = input_grad.value();
+        TT_FATAL(
+            tt::operations::primary::is_same_shape(input, input_grad_tensor), "both tensors should be the same shape");
+    }
+
+    if (weight_grad.has_value()) {
+        const auto& weight_grad_tensor = weight_grad.value();
+        TT_FATAL(
+            tt::operations::primary::is_same_shape(weight, weight_grad_tensor),
+            "both tensors should be the same shape");
+    }
+
+    if (bias_grad.has_value()) {
+        const auto& bias_grad_tensor = bias_grad.value();
+        TT_FATAL(
+            tt::operations::primary::is_scalar(bias_grad_tensor) ||
+                tt::operations::primary::is_1d_tensor(bias_grad_tensor),
+            "bias_grad tensor should be 1d or scalar");
+    }
+}
+
+std::vector<int64_t> find_reduce_dim(
+    const tt::tt_metal::LegacyShape& a_shape, const tt::tt_metal::LegacyShape& b_shape) {
+    std::vector<uint32_t> a_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    std::vector<uint32_t> b_dim(tt::tt_metal::MAX_NUM_DIMENSIONS, 1);
+    get_tensor_dim(a_dim, a_shape);
+    get_tensor_dim(b_dim, b_shape);
+    int32_t rank = std::max(a_shape.rank(), b_shape.rank());
+    log_debug(tt::LogOp, "find_reduce_dim :{} rank {} a {} b {}", __LINE__, rank, a_shape.rank(), b_shape.rank());
+    std::vector<int64_t> dims;
+    // batch dims
+    for (int i = 0; i < rank - 2; ++i) {
+        int idx = rank - 1 - i;
+        TT_ASSERT(idx >= 0);
```

Comment:
```suggestion
        TT_FATAL(idx >= 0);
```

Reply:
resolved in [9b40fca](https://github.com/tenstorrent/tt-metal/pull/12812/commits/9b40fca6fe2a092b7588eaeac2d24d998ecb517d)